walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,113 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> validTimes = new ArrayList<>();  // return of all valid ranges
+    List<TimeRange> invalidTimes = new ArrayList<>();  // keeps track of invalid ranges
+    Collection<String> attendees = request.getAttendees();
+
+    int duration = (int) request.getDuration();
+
+    // if length of meeting is longer than a whole day, return empty list
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+        return validTimes;
+    }
+
+    // if no conflicting events
+    if (events.isEmpty()) {
+        validTimes.add(TimeRange.WHOLE_DAY);
+        return validTimes;
+    }
+
+    // Add to invalid times collection
+    for (Event event : events) {
+        for(String attendee : attendees) {
+            if(event.getAttendees().contains(attendee)) {
+                invalidTimes.add(event.getWhen());
+            }
+        }
+    }
+
+    // if no events for attendees
+    if (invalidTimes.isEmpty()) {
+        validTimes.add(TimeRange.WHOLE_DAY);
+        return validTimes;
+    }
+
+    invalidTimes = mergeOverlap(invalidTimes);
```

Comment:
I might be completely stupid, seems to me that we don't need to merge overlap, and findValidTimes should also work just fine?

Reply:
When I get rid of the merge, most of my optional attendee tests don't pass and the nestedEvents test doesn't pass. The issue is with nested events and the last if check. If the last event is nested and shorter than the event before it, start will equate itself with the shorter event's end, giving and incorrect range. If I move start to the bottom, then it'll break other conditions.

Day:       |-------------|
Events:       |-------|
                     |---|
Meeting: |---|
                          |---| <-- wrong range

Reply:
Added a new variable to account for this. No more merge method!

Reply:
hmmm I see, you're right that wouldn't work. It might work if you keep a latestEndTime and don't add any valid time range if the current invalidTime ends before your latestEndTime. Merging also works so it's up to you.